Drug labels in pml

c-brenn opened the pull request:
new drug format as per customer's requests.

shawa committed:
Draft new API reference

shawa committed:
Draft new API reference

shawa committed:
Add labels parameter to drugs dinto query, to filter by name

shawa committed:
Add labels parameter to drugs dinto query, to filter by name

shawa committed:
Add uris endpoint

shawa committed:
Add uris endpoint

shawa committed:
Find DDIs by URL

shawa committed:
Find DDIs by URL

c-brenn committed:
asclepius controller

* removes ddi fetching from pml controller
* adds endpoint for fetching ddis for a given list of dinto uris
* adds endpoint for fetching dinto uris for a given list of drugs

c-brenn committed:
new drug format

drugs are now specified in PML using `drug { "..." }`.

c-brenn committed:
make frontend work

ugly code but works :older-woman:

c-brenn committed:
rebase

c-brenn commented:
drugs can be specified in PML like so
```
requires { drug { "paracetamol" } }
```

* `/api/pml` returns the list of english drugs in a pml file
* `/api/uris` takes a list of english drugs and uses the new asclpius endpoints to find their uris
* `api/ddis` takes a list of uris and returns their ddis

The front-end code works but is nice and ugly. The `content-type` has to be explicitly set to `application/json` or else phoenix thinks it's `text/plain`

c-brenn commented on panacea/web/static/js/app.js:
```diff
@@ -12,17 +12,48 @@ async function submitFile() {
+      if (uris_response.ok) {
+        const data = await uris_response.json();
+        const uris = Object.keys(data.uris.found);
+        console.log(uris);
```
will remove this

c-brenn committed:
removes console.log

houli commented on asclepius/README.md:
````diff
@@ -82,28 +82,57 @@ Endpoints
+##### Request Body
+
+```json
+{"labels": [ "paracetamol", "flat seven up", "cocaine"]}
````
Does this need to be labels or can we just send the list with no object around it?

houli commented on asclepius/README.md:
````diff
@@ -82,28 +82,57 @@ Endpoints
 ```json
-{"drugs": ["chebi:421707", "chebi:465284", "dinto:DB00503", "chebi:9342"]}
+{
+  "drugs": [
````
Does this need to be drugs or can we just send the list with no object around it? Realise this is an old one but sure

c-brenn replied:
Perhaps not needed - I kind of like the context that the outer object gives though.

houli commented on asclepius/README.md:
````diff
@@ -82,28 +82,57 @@ Endpoints
+
+```json
+{
+  "found": {
````
Maybe we want this instead
```json
{
  "found": [
    {
      "uri": "http://purl.obolibrary.org/obo/CHEBI_27958",
      "label": "cocaine"
    },
    {
      "uri": "http://purl.obolibrary.org/obo/CHEBI_46195",
      "label": "paracetamol"
    }
  ],
  "not_found": [
    "flat seven up"
  ]
}
```

c-brenn replied:
our reasoning for the current format is that when DDIs are returned the front-end will need to lookup the names of the drugs in each DDI. The DDIs return URIs for the participating drugs. The map format makes this easy.

houli commented on asclepius/asclepius/dinto.py:
```diff
@@ -57,14 +57,25 @@ def sparqled(*args, **kwargs):
+def drugs(labels=None):
+
+    if labels is not None:
+        if not isinstance(labels, frozenset):
```
Not sure we need to catch this, we should accept any hashable, if it's not hashable then the lru_cache will throw anyway

houli commented on asclepius/asclepius/dinto.py:
```diff
@@ -81,22 +92,16 @@ def all_ddis():
     if not isinstance(drugs, frozenset):
         raise ValueError("for cachability, `drugs` must be given as a frozenset")
-
+    print(drugs)
```
Presume this isn't needed

houli commented on panacea/web/static/js/app.js:
```diff
@@ -12,17 +12,47 @@ async function submitFile() {
   hideResults();
   try {
-    const response = await fetch(this.action, {
+    const drugs_response = await fetch(this.action, {
```
A style thing for the JS bits camelCase it

shawa committed:
Merge branch 'drug-labels-in-pml' of github.com:tom-and-the-toothfairies/pathways into drug-labels-in-pml

houli commented on panacea/web/static/js/app.js:
```diff
@@ -12,17 +12,47 @@ async function submitFile() {
       method: 'POST',
       body: new FormData(this),
       credentials: 'same-origin',
       headers: new Headers({authorization: apiAccessToken})
```
We'll probably want `'Accept': 'application/json, text/plain, */*'` headers on all of these requests

c-brenn committed:
headers and camel case

shawa committed:
Merge branch 'drug-labels-in-pml' of github.com:tom-and-the-toothfairies/pathways into drug-labels-in-pml

c-brenn committed:
updates documentation

22a commented:
Working on the live updates(drugs show up before ddis are queried), pls don't merge yet 💇‍♂️

22a committed:
added live UI updating to the request chain :link:

22a committed:
fixed merge conflict between :camel:s

c-brenn commented on asclepius/asclepius/main.py:
```diff
@@ -63,15 +63,31 @@ def ddis():
+    """
+    params = request.get_json()
+    if params is None or 'labels' not in params or not params['labels']:
+        raise InvalidUsage("Expecting {'labels': [...]} with at least two labels")
```
this message doesn't really match up

shawa committed:
Remove _valid_drug test

shawa committed:
Merge branch 'drug-labels-in-pml' of github.com:tom-and-the-toothfairies/pathways into drug-labels-in-pml

houli committed:
Fix python tests

22a committed:
added conditional breaks in request chain

added display panel for unidentified drugs
removed superfluous class addition to panels

22a committed:
Merge branch 'drug-labels-in-pml' of github.com:tom-and-the-toothfairies/pathways into drug-labels-in-pml

c-brenn commented on panacea/web/static/js/app.js:
```diff
@@ -27,35 +27,49 @@ async function submitFile() {
+        if (urisResponse.ok) {
+          const data = await urisResponse.json();
+          const uris = Object.keys(data.uris.found);
+          const unidentifiedDrugs = Object.values(data.uris.not_found);
```
not sure if you need Object.values here.  I think not_found is just a list.

22a replied:
oops lol

22a commented:
few things going on in that last commit:
* added unidentified drugs panel
* added sample pml file with silly "drugs" for manual visual testing of panels
* removed superfluous class removal in the display functions (the classes get reset at the start of the request chain)

the unidentified panel looks like this:
<img width="1393" alt="screen shot 2017-03-08 at 21 37 27" src="https://cloud.githubusercontent.com/assets/7144173/23725639/45bcc6f6-0449-11e7-84fb-dd930d659637.png">


it also raises a few questions:
* what do we want to display to the user when the request chain is broken at the different points (when found URI's is less than `2` and when identifiable drugs is less than `0`{thinking about this, this should also probably be `2`, as there's no point getting the uri of one drug alone}) 
* do we want to make the error panel more expressive (where we can change the panel header as well)

22a commented:
also, the request chain is getting quite hairy now(v nested), perhaps we could consider breaking the chain out to separate handlers? **bleugh**

c-brenn commented:
Nice work.

👍 yeah I'm all for refactoring the request chain.

I think we need a bigger discussion about the UI that's beyond the scope of this PR.

22a commented on panacea/web/static/js/app.js:
```diff
@@ -27,35 +27,49 @@ async function submitFile() {
-        displayError(await urisResponse.json());
+      }
+      else {
+        displayError({"message": "No drugs found"});
```
this is old syntax, `{message: "bleugh"}` would be nicer, i see we have used it above

22a committed:
minor style changes to js

22a committed:
missed an 'else' :goat:

c-brenn committed:
clarified asclepius/uris error message

c-brenn commented:
think this is ready to go - any further changes needed?

22a commented:
Looks good to 🚢  the 🐶 to me. @houli will need to approve his requested changes as I feel dismissing his review would be a breach of interpersonal trust and set a worrying precedent

c-brenn commented:
Actually, the frontend changes should go in he change log.

* the unidentified drugs
* change in spinner behaviour

22a commented:
Good spot, will add those now

22a committed:
added request chain UI info to changelog

c-brenn committed:
changes format of asclpius/uris response

`found` is now a list of objects instead of one big map
